### ISSUE TITLE
Making two LSTMs each using different input subsequence

### ISSUE THREAD

**Issue by lechuzo** [fchollet/keras#3145]
hi,

I use LSTM for text classification (return_sequences=False).
Each input instance is n size words. Now instead of feeding the whole input sequence, I'd like to feed the first m (m<n) words into LSTM and feed a second LSTM with m+1..n words (the rest of the sequence) and concatenate the vectors of each LSTM output (and use a dense layer after/before concatenation)

It that possible with Keras? if not, should I wrap the LSTM class with my own? how?

thanks a lot

![lstmx2](https://cloud.githubusercontent.com/assets/3630332/16577891/640f95cc-42a0-11e6-9225-ad2c66599052.jpg)

**Comment by a-rodin** [comment]
You can use Lambda layer to cut first m symbols for the first LSTM and the rest of the symbols for the second and then Merge layer for models with each of these LSTMs.

Something like this:
```python
from keras.models import Sequential
from keras.layers import Lambda, Merge, Activation, Dense
from keras.layers.recurrent import LSTM

n, m = 10, 5
input_dim = 10
output_dim = 20

first_model = Sequential()
first_model.add(Lambda(lambda x: x[:,:m,:], input_shape=(n, input_dim)))
first_model.add(LSTM(output_dim))

second_model = Sequential()
second_model.add(Lambda(lambda x: x[:,m:,:], input_shape=(n, input_dim)))
second_model.add(LSTM(output_dim))

model = Sequential()
model.add(Merge([first_model, second_model], mode='concat', concat_axis=1))
model.add(Dense(1))
model.add(Activation('sigmoid'))
model.compile(optimizer='RMSprop', loss='binary_crossentropy')
```

**Comment by lechuzo** [comment]
looks elegant. thanks

So I can work with multiple models, and thus determine the input of each.
How about the case where the input sequence is actually two sequences (ie m varies among sentences). In fact input x has 2 lists of words.

thanks a lot @a-rodin 

minor thing - maybe a Dense layer could be useful after each LSTM

**Comment by a-rodin** [comment]
@lechuzo If m varies among sentences how is it determined? Is it pre-computed on stage of preparing the dataset, or is it should be computed inside of the model in dependence of some model's weights?

**Comment by lechuzo** [comment]
it's precomputed.
say x (data input) has 2 lists:
x['words1]=[w1, w2, w3]
x['words2]=[w4, w5, w6]

think of words1 and words2 as 2 different sentences

**Comment by a-rodin** [comment]
So you surely can create two sequences padded to zeros and feed them both to `model.fit` instead of `[X[:,:m,:], X[:,m:,:]]`.

Btw do you mind about using embedding before LSTM?

**Comment by lechuzo** [comment]
thanks.
sounds cool. Did I get it ? (notice the 'validation_data' also)


X1, X2=[]
for item in X:
    X1.append(item ['words1'])
    X2.append(item ['words2'])

X_test1, X_test2=[]
for item in X_test:
    X_test1.append(item ['words1'])
    X_test2.append(item ['words2'])

first_model = Sequential()
first_model.add(Embedding(output_dim=w2v_dim, input_dim=max_features, mask_zero=True,  input_length=maxlen, dropout=0.2))
first_model.add(LSTM(output_dim)
first_model.add(Dense(1))

second_model = Sequential()
second_model.add(Embedding(output_dim=w2v_dim, input_dim=max_features, mask_zero=True,  input_length=maxlen, dropout=0.2))
second_model.add(LSTM(output_dim, input_shape=(n-m, input_dim)))
second_model.add(Dense(1))

model = Sequential()
model.add(Merge([first_model, second_model], mode='concat'))
    # optional instead\in addition to Dense in the 2 sub-models : model.add(Dense(1))
model.add(Activation('softmax')) 

model.compile(optimizer='RMSprop', loss='binary_crossentropy')
model.fit([X1, X2], y,  validation_data=([X_test1, X_test2], Y_test))

**Comment by a-rodin** [comment]
Looks nice, but you haven't pad sequences. It can be done with keras function `pad_sequences`. Check `babi_rnn.py` example in keras demos to check out how does it work.

**Comment by a-rodin** [comment]
@lechuzo You haven't pad sequences. It can be done with keras function `pad_sequences`. Check `babi_rnn.py` in keras demos to see example usage of it.

**Comment by wailoktam** [comment]
@lechuzo 
Thanks. I am thinking of using embedding layer as well. So the part of speech types are the lexical entries in the vocabulary. 

I think the POS and the actual words should each work like a color channel in image. Some friends suggest I can merge them by the concat mode. I wonder whether it is the right thing to do, particularly after reading your original post.

You merge the output of w1 w2 w3  with the output of  w4 w5 w6  by concatenation. w1 and w4 are information at separate locations of the whole input.

So creating two input sequences (or making embedding of the two sequences) w1 w2 w3 and p1 p2 p3 would mean information at separate location of the whole output, I think? This does not feel like what I want. w1 relates to p1 just like the green channel value relates to the red channel value at a point in an image.  

Thanks in advance for any help.

**Comment by vinayakumarr** [comment]
Any body having character based text classification code in keras like in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/skflow/text_classification_character_rnn.py

**Comment by abhilash1512** [comment]
how can i implement the merging in tensorflow?
